Add a WildcardAssembler Plugin.

plucia-mitre opened the pull request:
Allows assembling instructions containing wildcards. Includes support for filtering assembly results to only include some completions for each wildcard.

This starts to address issue #5800 which may also address several other existing issues including #679 and #2516.

This pull request builds on the assembler changes merged in e7458ed and was developed alongside @nsadeveloper789 (to whom we are very thankful for his support!)

This WildcardAssembler is not specific to Pickled Canary (though certainly useful for it). In addition to Pickled Canary and other assembly search tools, this assembler would be useful for tools which could now more easily surface architectural details about assembly instructions to end-users or other tools.

For more background on this pull request, see the now-closed #5801

plucia-mitre commented on Ghidra/Features/WildcardAssembler/src/main/help/help/TOC_Source.xml:
It's unclear to me if I need to add something else to this file to make the documentation work properly.

plucia-mitre replied:
Thanks! I'll do some combo of the above and we can edit from there :-)

plucia-mitre replied:
I've added two example scripts and tweaked the documentation. I've verified that the documentation looks as expected in the generated help (Help > Contents).

I certainly could be missing something, but the `gradle WildcardAssembler:tidyHelp` gradle command gives me:

```
Cannot locate tasks that match 'WildcardAssembler:tidyHelp' as task 'tidyHelp' not found in project ':WildcardAssembler'.
```

Furthermore, I'm not able to find any references to 'tidyHelp' in the repo unless GitHub search is letting me down! Thanks for your ongoing help!

dragonmacher replied:
The `tidyHelp` task is simply a call to an HTML beautification library.  It appears as though it is not in the repo, but instead inside of a build file that we use internally.    It is not necessary for you to run this task.

plucia-mitre replied:
That makes sense; thanks for letting me know I'm not *completely* crazy! 😄

plucia-mitre commented:
For reference (though certainly not critical to this request), there's a pull request for Pickled Canary here (mitre/pickled-canary#3) which refactors Pickled Canary to use this Wildcard Assembler.

nsadeveloper789 commented on Ghidra/Features/WildcardAssembler/ghidra_scripts/FindInstructionWithWildcard.java:
```diff
@@ -0,0 +1,346 @@
+	 * {@code WildAssemblyResolvedPatterns}
+	 * 
+	 * @param results
+	 * @return
```
Missing descriptions for param and return

nsadeveloper789 commented on Ghidra/Features/WildcardAssembler/ghidra_scripts/WildSleighAssemblerInfo.java:
```diff
@@ -0,0 +1,157 @@
+		// Make sure that if one of the example instructions was chosen the current binary has the
+		// correct architecture.
+		if (sampleInstructions.contains(wildcardedInstruction) &&
+			!language.getLanguageID().toString().equals("x86:LE:64:default")) {
```
Ditto from the other script.

plucia-mitre replied:
This is different from the other script because here I am just raising a warning if the user is using one of the example instructions with an unexpected architecture). I've removed the "return" statement right after this to allow us to try anyways (and updated the warning text slightly). Do you think I should do something different here?

nsadeveloper789 replied:
Oh. Honestly, I missed the `if` statement on Line 75. I'll re-read as I check your other revisions.

nsadeveloper789 commented on Ghidra/Features/WildcardAssembler/ghidra_scripts/FindInstructionWithWildcard.java:
Nice script. Even though it's a demo, for something provided out of the box, we should strive to make it as generally useful as possible without requiring the user to modify it. I marked particulars below as I was reading through it, but some usablility tips:

* Just use the current program and its language. The assembler is supposed to accommodate all Ghidra's languages, anyway (even though we know it's not perfect.)
* Allow the user to input the instruction. I see you use `askString()` in one script but not the other....
* I had suggested using a table, but IIRC, addresses become links in the script Console? If so, then don't worry about it. Otherwise, consider just finding the next result and calling `goTo`, or using the table.

Ideally, the user would only need to modify the script to add a feature or change its operation, not just as a means of providing input.

nsadeveloper789 reviewed (changes requested):
The scripts and the help are a nice improvement! Thank you. Many coding style comments, then a few usability comments for the scripts. These aren't hard a fast "you musts," but strong suggestions. If you've already tried some of the suggestions and ran into issues, let me know.

nsadeveloper789 commented on Ghidra/Features/WildcardAssembler/src/main/help/help/topics/WildcardAssemblerPlugin/Wildcard_Assembler.html:
```diff
@@ -0,0 +1,94 @@
+      <CODE>!</CODE> character. For example, in a x86:LE:32:default binary:</P>
+      <BLOCKQUOTE>
+        <DL>
+          <DT>Assembles (no wildcard):</DT>
```
Probably remove the "Assembles" and parentheses for these. Just, e.g.:

> No wildcard:

For the examples that doesn't work:

> Single token (does *not* assemble):

plucia-mitre commented on Ghidra/Features/WildcardAssembler/ghidra_scripts/FindInstructionWithWildcard.java:
```diff
@@ -0,0 +1,346 @@
+
+		Memory memory = currentProgram.getMemory();
+
+		for (var entry : encodings.entrySet()) {
```
I'm resetting the `matchCount` inside the outer loop (and the `searchFromAddress`) so I'll get up to 10 results for each of the encodings (rather than 10 total). There are certainly other ways to do it (and I'm happy to consider other suggestions) but my thought was to emphasize the different encodings in this demo as that's the "special sauce" of the WildcardAssembler and one of the concepts a dev needs to understand to use it. As a search tool, yes, having the search results in order regardless of encoding probably makes more sense, but my thought is the point of this script is more to show how to use the WildcardAssembler so other folks can make cool stuff. The great search tool is coming along with the other components in subsequent pull requests / Pickled Canary versions. Thoughts?

plucia-mitre commented on Ghidra/Features/WildcardAssembler/ghidra_scripts/FindInstructionWithWildcard.java:
```diff
@@ -0,0 +1,346 @@
+				// Get the specific bytes found at this address and print match info
+				var foundBytes = new byte[encoding.length()];
+				memory.getBytes(matchAddress, foundBytes);
+				printSearchHitInfo(matchAddress, foundBytes, allValidResolvedPatterns);
```
Correct me if I'm wrong, but I'm not sure these give me a great way to also print the wildcard information for each hit.

plucia-mitre commented on Ghidra/Features/WildcardAssembler/ghidra_scripts/FindInstructionWithWildcard.java:
```diff
@@ -0,0 +1,346 @@
+					return false;
+				}
+			}
+			return true;
```
I've reworked this a little bit and added comments. I'm fairly sure we can't use a map based on wildcard (name) because an instruction could have multiple wildcards in it with the same name.

plucia-mitre commented:
I think I've addressed most of the above feedback but I've left a few comments on several of the points. Let me know if you'd like me to mark the other threads as resolved or leave them open for someone else to confirm my fixes.

nsadeveloper789 reviewed (commented):
This looks good. Thanks for the revisions! At this point, I think any remaining revisions should be minor, and we can take care of them during our internal review process. No guarantees, but we should be good to go.

plucia-mitre commented:
Great! Thanks! Let me know if there's anything else here, else I'll be working on cleaning up the search engine and pattern parser portions of Pickled Canary as time permits.

astrelsky commented:
![8f0cos.jpg](https://github.com/NationalSecurityAgency/ghidra/assets/46897303/5a2c7906-e0be-4635-b530-793b3c6fd3cc)

I couldn't help myself, someone had to do it.
